components/bl-audio-recorder-component/src/index.js

```diff
@@ -0,0 +1,155 @@
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+
+import { captureMediaDevices, download } from './helpers';
+
+const { cn, normalizeDimensionValue } = BackendlessUI.CSSUtils;
+
+export default function AudioRecorder({ component, eventHandlers, elRef }) {
+  const {
+    player, controls, noise, fileName, fileType, width, startText, stopText, downloadText, pauseText, resumeText,
+    display, style, classList,
+  } = component;
+  const { onStart, onStop, onDownload, onState } = eventHandlers;
+
+  const audioRef = useRef();
+  const recorderRef = useRef();
+
+  const [recordedBlob, setRecordedBlob] = useState();
+  const [state, setState] = useState();
+
+  const styles = useMemo(() => ({
+    width: normalizeDimensionValue(width),
+    ...style,
+  }), [style]);
+
+  useEffect(() => {
+    if (recordedBlob && audioRef.current) {
+      audioRef.current.src = URL.createObjectURL(recordedBlob);
+    }
+  }, [recordedBlob]);
+
+  useEffect(() => {
+    if (state) {
+      onState({ state });
+    }
+  }, [state]);
+
+  Object.assign(component, {
+    start       : () => startRecording(),
+    stop        : () => stopRecording(),
+    downloadFile: () => downloadRecordedFile(),
+    getBlob     : () => recordedBlob,
+    pause       : () => toggleRecord(),
+  });
+
+  const startRecording = useCallback(async () => {
+    const audioConstrains = { audio: { echoCancelation: true, noiseSuppression: noise } };
+    const stream = await captureMediaDevices(audioConstrains, audioRef);
+
+    if (stream) {
+      recorderRef.current = new MediaRecorder(stream);
+
+      const chunks = [];
+
+      recorderRef.current.ondataavailable = event => {
+        if (event.data.size > 0) {
+          chunks.push(event.data);
+        }
+      };
+
+      recorderRef.current.onstart = () => {
+        setState(recorderRef.current.state);
+      };
```

Comment:
```
const stateSetter = () => setState(recorderRef.current.state);

recorderRef.current.onstart = stateSetter
recorderRef.current.onpause = stateSetter
recorderRef.current.onresume = stateSetter
```

or something like
``` 
recorderRef.current.onstart = recorderRef.current.onpause = recorderRef.current.onresume = stateSetter
```

or

```
Object.assign( recorderRef.current, {
onstart: stateSetter,
onpause: stateSetter,
onresume: stateSetter,
})
```

Also we could make it obvious if you know the state using:
```
setState(Stream.State.PAUSED)
```

and so on